feat: add preliminary Windows support and mruby_dir feature

2moe opened the pull request:
1. Adding a `Default` implementation for `MRubyManager` was suggested by `cargo clippy`. Honestly, I recommend running `cargo clippy` manually. 

2.
Some documentation comments are automatically wrapped to the next line due to excessive width. This does not affect the final display. For more precise formatting, please provide a `.rustfmt.toml` file in the project root.

3.
- old: `self.mruby_version.map(String::from)`
- new: `self.mruby_version`

Since mruby_version is already of type `Option<String>`, using map here is unnecessary.

4.

`return Err(anyhow!(format!("{fmt}")))` => `bail!("{fmt}")`

It can be simplified.  
 
This is the basic API usage of `anyhow`.
I'm surprised you don't know it.

Close #24

2moe committed:
feat(build-utils): add preliminary support for Windows

2moe committed:
feat(minutus): add "mruby_dir" feature

2moe commented:
update: 
  feat(minutus): add "mruby_dir" feature

We can do this:

```sh
# POSIX-sh
#
env MINUTUS_MRUBY_DIR=/tmp/mruby-local-dir cargo b -rvv
```

2moe committed:
chore(minutus/build.rs): rename abs path

2moe committed:
chore(minutus/build.rs::copy_to_mruby_outdir): update warn msg

2moe committed:
feat(minutus): add "dir_is_not_empty" function

2moe committed:
docs(readme): cross-platform => Supported Platforms

genya0407 commented on README.md:
```diff
@@ -194,6 +194,8 @@ Following versions are supported:
 
 You can also use mruby's `master` branch, but it is not tested.
 
+If you need to customize a specific version (such as a specific tag), you can do so via `mruby_dir`. See "Local Dir" for details.
```
[nits]

```suggestion
If you need to use mruby versions not listed above (e.g., a specific tag or a fork), you can do so via `mruby_dir` feature. See "mruby_dir" section for details.
```

genya0407 commented on README.md:
```diff
@@ -209,6 +211,60 @@ minutus = { version = "*", features = ["mruby_3_2_0"] }
 
+### Local Dir
+
+Use a local directory instead of a remote Git repository.
```
[nits]

```suggestion
### mruby_dir

When `mruby_dir` feature is specified, minutus uses your local directory as `mruby`.
```

genya0407 commented on README.md:
````diff
@@ -209,6 +211,60 @@ minutus = { version = "*", features = ["mruby_3_2_0"] }
+[dependencies]
+minutus = { version = "*", features = ["mruby_dir"] }
+```
+
````
We had better describe `MINUTUS_MRUBY_DIR` here, like:

````suggestion
```sh
# POSIX-sh
#
env MINUTUS_MRUBY_DIR=/path/to/my/mruby cargo build
```
````

2moe replied:
After some consideration, I decided to configure the environment variables in the **Rakefile** instead of using POSIX sh. 

```ruby
# The value of `MINUTUS_MRUBY_DIR` is the path to the source code of mruby.
ENV["minutus_mruby_dir".upcase] = "/tmp/mruby-stable"
```

I chose to use `'minutus_mruby_dir'.upcase` rather than `MINUTUS_MRUBY_DIR`, as I personally find it more readable and expressive.

It might be better to update minutus-mrbgem-template directly, rather than relying on users to follow the instructions in the README manually.

@genya0407 As the maintainer of this project, you can improve it yourself later.

genya0407 replied:
Writing the environment variable in Rakefile in Example section is good.
But, I want to describe `MINUTUS_MRUBY_DIR` **before** example section.

Reason 1:
Generally speaking, I would like to provide a complete description before the example.
I think examples are meant to help the user understand the feature or describe additional uses, and so on.

Reason 2:
minutus has two usecases:

1. Embed mruby code in Rust code
    * https://github.com/genya0407/minutus/tree/trunk?tab=readme-ov-file#embed-mruby-in-rust
2. Create mrbgem by Rust
    * https://github.com/genya0407/minutus/tree/trunk?tab=readme-ov-file#create-mrbgem-by-rust

`mruby_dir` is for both usecases, however, in former one, users do not necessarily have `Rakefile`.

---

If you do not have strong opinions about the README, could I rewrite it after merging this PR?
After my rewrite, if you happen to have any concerns about the README, please feel free to propose any changes.

Of course, if you want to continue discussion, that's welcome!

2moe replied:
I have no objections to this.

genya0407 commented on README.md:
```diff
@@ -209,6 +211,60 @@ minutus = { version = "*", features = ["mruby_3_2_0"] }
+
+**minutus** primarily supports Unix-like systems.
+
+If you encounter issues in a Windows MSVC environment, try using **msys2** and the `*-pc-windows-gnu` target.
```
👍

genya0407 commented on README.md:
````diff
@@ -209,6 +211,60 @@ minutus = { version = "*", features = ["mruby_3_2_0"] }
+```sh
+# POSIX-sh
+#
+env MINUTUS_MRUBY_DIR=/tmp/mruby-local-dir cargo b -rvv
````
Why not use the "canonical form" of command? (`cargo build`, not `cargo b`)
And I think we shouldn't add extra options unless it's necessary to avoid confusion for users.

```suggestion
env MINUTUS_MRUBY_DIR=/tmp/mruby-local-dir cargo build
```

2moe replied:
Sorry, that was my oversight. 
When I wrote the README, I didn’t think too much about it—it just came out naturally.

You can use `cargo --list` to see all available Cargo subcommands. Although `b` is a built-in alias provided by Cargo, I agree that `build` is more intuitive. 

The output of `cargo build -h` includes options like `-r, --release` and `-v, --verbose... Use verbose output (-vv very verbose/build.rs output)`. 

While `-rvv` might not be immediately intuitive, it's not hard to understand.
The reason for using `-vv` is to capture the output from `build.rs` during the build process, since errors can occur at that stage. Giving users more visibility into the build process is not necessarily a bad thing.

genya0407 commented on minutus-mruby-build-utils/src/lib.rs:
```diff
@@ -63,13 +75,10 @@ impl MRubyManager {
-        let mruby_version = self
-            .mruby_version
-            .map(String::from)
-            .expect("mruby_version is not set.");
```
Oh... I don't know why I did this...

genya0407 commented on minutus-mruby-build-utils/src/lib.rs:
```diff
@@ -86,16 +95,42 @@ impl MRubyManager {
+        #[cfg(not(windows))]
+        _ => "rake",
+    };
+
```
This is a bit comprecated.
Could we write this as follows?

```suggestion
    let rake = if (cfg!(windows)) {
      "rake.bat"
    } else {
      "rake"
    };
```

This change might add runtime overhead, but I think it's negligible.

2moe replied:
I updated it to `if cfg!(windows)` since the outer parentheses in `if (cfg!(windows))` are unnecessary.

genya0407 commented on minutus-mruby-build-utils/src/lib.rs:
```diff
@@ -86,16 +95,42 @@ impl MRubyManager {
+    #[allow(unreachable_patterns)]
+    let mruby_config = match mrb_cfg_bin {
+        #[cfg(windows)]
+        p => ["bat", "exe"]
```
Just curious: are there any situations where `mruby-config.bat` does not exist on Windows?

You wrote before:

>On Windows, "build/host/bin" does not have mruby-config, only mruby-config.bat`

https://github.com/genya0407/minutus/issues/24#:~:text=On%20Windows%2C%20%22build/host/bin%22%20does%20not%20have%20mruby%2Dconfig%2C%20only%20mruby%2Dconfig.bat

2moe replied:
To begin with: checking whether `mruby-config` exists is not a bad thing.

Previously, I tried to minimize the size of `mruby-bin-mruby` by trimming the default configuration. Although I forgot which platform it was, I vaguely remember that the generated `bin/` directory did not contain `mruby-config`. Of course, I might be mistaken, but I still don't think checking for the file's existence is unnecessary.

I first encountered this project last week, and I was using Windows at the time. Since the first step was to install `minutus-mrbgem-template`, and there was no prebuilt binary available, I had no choice but to compile it myself. Naturally, the experience was a bit unpleasant.

> Because the first step is already so hard, many people might give up right away.
> They probably don’t have as much patience as I do.

Your code uses a lot of `unwrap()`, which means that in many cases, I have to manually understand the build process to figure out what went wrong.

A error like "`mruby-config` does not exist" is better than a later `unwrap()`.

On Windows, a file extension is required for direct execution.
If `a.exe` exists, you can run it with `.\a` or `.\a.exe`.
But if you rename `a.exe` to `a`, then double-clicking the file or running `.\a` via conhost + cmd won’t work.
(Unless you're using a non-standard Windows shell environment.)

In theory, Rust’s `Command::new("a")` should be able to detect `a.bat`, since `.BAT` is part of `PATHEXT`.
In practice, it can detect `a.exe`, but not `a.bat`.

Using `["bat", "exe"].iter()` doesn’t add extra overhead, since `"bat"` has higher priority in the array and Rust’s iterators are lazy.

The Ruby in `msys2 ucrt64` indeed only provides ".bat"(`mruby-config.bat` and `rake.bat`).

However, checking both `.bat` and `.exe` is not always redundant.
When I encountered this issue, my first thought was to write an `.exe` wrapper.

Now, I’ve changed the code to `.bat` only, which might reduce the maintenance burden.

genya0407 commented on minutus-mruby-build-utils/src/lib.rs:
```diff
@@ -133,31 +166,29 @@ pub fn download_mruby(workdir: &Path, mruby_version: &str) {
 
 fn run_command(current_dir: &Path, cmd: &[&str]) -> Result<String> {
-    println!("Start: {:?}", cmd);
+    println!("Start: {cmd:?}");
```
Pretty nice!

genya0407 commented on minutus/build.rs:
```diff
@@ -74,16 +73,28 @@ fn main() -> Result<()> {
     )?;
 
     let do_link = env::var("CARGO_FEATURE_LINK_MRUBY").is_ok();
+    let do_download = match env::var("CARGO_FEATURE_MRUBY_DIR") {
```
I think `do_download` and `mruby_dir` cannot be used together, because minutus cannot determine the download target when `mruby_dir` is enabled.
So we should set `do_download = false` if `CARGO_FEATURE_MRUBY_DIR` is set.

2moe replied:
Simply put, I originally supported both `mruby_dir` and `mruby_[version]` out of fallback considerations.  
The code has now been modified, and I’ll explain it later.

genya0407 commented on minutus/build.rs:
```diff
@@ -74,16 +73,28 @@ fn main() -> Result<()> {
+    let do_download = match env::var("CARGO_FEATURE_MRUBY_DIR") {
+        Err(_) => true,
+        _ => match env::var("MINUTUS_MRUBY_DIR") {
+            Ok(dir) if dir.trim().is_empty() => true,
```
I cannot imagine `mruby_dir` usecases where `MINUTUS_MRUBY_DIR` is not set (or empty string is set).
Why not `bail!` if `mruby_dir` is enabled but `MINUTUS_MRUBY_DIR` is not set?

2moe replied:
As with the above, this is due to fallback considerations.
I’ll explain it later.

genya0407 commented on minutus/build.rs:
```diff
@@ -92,28 +103,76 @@ fn main() -> Result<()> {
+        .is_some()
+}
+
+fn copy_to_mruby_outdir(
```
Maybe this function should be located inside `MRubyManager`.
Something like:

```rust
MRubyManager::new()
    .copy_mruby_from("path/to/mruby")
    .run();
```

2moe replied:
OK.

```rust
pub struct MRubyManager {
    // ...
    copy_mruby_from: Option<PathBuf>,
}
impl MRubyManager {
    pub fn copy_mruby_from<P: Into<PathBuf>>(mut self, path: P) -> Self {
        self.copy_mruby_from = Some(path.into());
        self
    }
}
```

genya0407 commented on minutus/build.rs:
```diff
@@ -74,16 +73,28 @@ fn main() -> Result<()> {
         .link(do_link)
         .build_config(&build_config_copy)
+        .download(do_download)
         .run();
```
Furthermore, I think it's natural to run completely different logic depending on `CARGO_FEATURE_MRUBY_DIR`.

```rust
let do_link = env::var("CARGO_FEATURE_LINK_MRUBY").is_ok();
if env::var("CARGO_FEATURE_MRUBY_DIR").is_ok() {
  let mruby_dir = env::var("MINUTUS_MRUBY_DIR")?;
  MRubyManager::new() // We cannot provide proper `mruby_version` if mruby_dir feature is used.
    .link(do_link)
    .build_config(&build_config_copy)
    .download(false)
    .copy_mruby_from(mruby_dir)
    .run();
} else {
  // existing logics here...
  MRubyManager::new()
    .mruby_version(&mruby_version())
    .link(do_link)
    .build_config(&build_config_copy)
    .run();
}
compile_bridge();
```

2moe replied:
My idea is to avoid writing repetitive parts, keep it DRY (i.e., Don't Repeat Yourself)

```rust
    let mut manager = MRubyManager::new() 
    // We cannot provide proper `mruby_version` if mruby_dir feature is used.
        .link(do_link)
        .build_config(build_config_copy);

    manager = if env::var("CARGO_FEATURE_MRUBY_DIR").is_ok() {
        let mruby_dir = env::var("MINUTUS_MRUBY_DIR")?;
        manager
            .download(false)
            .copy_mruby_from(mruby_dir)
    } else {
        manager.mruby_version(&mruby_version())
    };
    manager.run();
    compile_bridge();
```

Actually, the above style is still written to suit your preferences. Personally, I would use a different style.

genya0407 reviewed (commented):
Thank you for your contribution ❤️

2moe committed:
feat(build-utils): add MRubyManager.copy_mruby_from &  mruby_dir mod

2moe committed:
docs(build-utils/mruby_dir::copy_to_mruby_dir):  clarify `cp -r` behavior in UNIX-like systems

2moe committed:
docs(readme): mruby-master => mruby-src-dir

2moe commented:
First, I’ve followed your suggestions: the `mruby_dir` and `mruby_[versions]` features are now mutually exclusive, and the code has been updated to use `copy_mruby_from(mruby_dir)`.

For the code you haven’t reviewed yet (`minutus-mruby-build-utils/src/mruby_dir.rs`), I’ve added extensive comments to help you understand why I wrote it that way.

Next, I’ll explain the previous implementation.

Although the original code has been modified, I still want to clarify why the original fallback strategy made sense:
`mruby_dir` and `mruby_[versions]` could coexist, and when `mruby_dir` failed, it would automatically fall back to `mruby_[versions]`.

Before diving deeper, let’s review some Rust fundamentals.
Rust has two basic concepts for composite data types:

- **Product**: Represents a “combination,” meaning a value contains multiple parts (fields).
  - We can loosely map this to `struct` or tuple.
- **Sum**: Represents a “choice,” meaning a value can only be one of several possible types.
  - We can loosely map this to `enum`.

Since an `enum` can contain a `struct`, and a `struct` can contain an `enum`, these two concepts can combine to form very complex type systems.
Therefore, this simple mapping isn’t entirely accurate—but as long as you get the idea, that’s fine.

Example:

```rust
enum Option<T> {
    Some(T),
    None,
}
```

I understand that what you want to express is that two similar features should be mutually exclusive—like a value `a: Option<u8>`.
`a` is either `Some(u8)` or `None`; it cannot be both at the same time.

Similarly, `feat: Features` should be either `Version` or `Dir`:

```rust
enum Features {
    Version("master"),
    Dir("/path/to/mruby"),
}
```

In reality, Cargo features behave more like a `struct` than an `enum`:

```rust
struct Features {
    master: true,
    v3_3_0: true,
    dir: true,
}
```

This means you can enable multiple features simultaneously; they combine and coexist.

When multiple similar features coexist, you need to consider **priority**.

In fact, this situation—multiple similar features coexisting—is not uncommon in popular crates.

For example, if you enable features `a`, `b`, and `c`:

- `a`, `b`, and `c` are similar.
- `a` has higher priority.
- If `a` fails to build, automatically fall back to `b`.

This approach exists because developers need to consider library usability and compatibility. It’s not unusual.

Of course, there are also scenarios where similar features are mutually exclusive:

```rust
#[cfg(all(feature = "a", feature = "b"))]
compile_error!("Features `a` and `b` are mutually exclusive");
```

Given that Cargo features are inherently combinable, I couldn’t guess your intention when submitting the PR.
Sorry—that’s my mistake.

Based on your suggestion, I modified `MRubyManager`:

```rust
pub struct MRubyManager {
    workdir: Option<PathBuf>,
    mruby_version: Option<String>,
    do_link: bool,
    build_config: Option<PathBuf>,
    do_download: bool,
    copy_mruby_from: Option<PathBuf>,
}
```

Since `do_download` and `copy_mruby_from` are mutually exclusive, although we can currently distinguish them with simple condition checks, I don’t think that’s the best solution.

I believe we should express this mutual exclusivity more clearly at the **type level**, or at least semantically:

```rust
// MRuby source code can only come from a remote location (requires download) or a local directory.
// It cannot come from both at the same time.
enum MRubySource {
    // Remote stores mruby_version. If you adjust build_on_doc_rs() to use a local path,
    // you can eliminate the `do_download` field.
    Remote(Option<String>),
    Local(Option<PathBuf>),
}
```

Since our approaches to Rust implementation differ, I won’t continue submitting code for "that sum type".

genya0407 commented on minutus-mruby-build-utils/src/lib.rs:
```diff
@@ -62,21 +90,29 @@ impl MRubyManager {
+            let mruby_version = self
+                .mruby_version
+                .expect("mruby_version is not set.");
+
```
nice

genya0407 commented on minutus-mruby-build-utils/src/lib.rs:
```diff
@@ -86,23 +122,44 @@ impl MRubyManager {
+        p if p.exists() => Some(p),
+        _ => None,
+    }
+    .expect(r#"The `mruby-config` executable file does not exist!"#);
```
Could we write this as follows?

```suggestion
    let mruby_config = if cfg!(windows) {  mrb_cfg_bin.with_extension("bat") } else { mrb_cfg_bin }
    if !mruby_config.exists() {
      panic!(r#"The `mruby-config` executable file does not exist!"#);
    }
```

genya0407 commented on minutus-mruby-build-utils/src/mruby_dir.rs:
````diff
@@ -0,0 +1,88 @@
+/// // src_dir => "{work_dir}/mruby"
+/// copy_to_mruby_dir(src_dir, &work_dir)?;
+/// ```
+pub(crate) fn copy_to_mruby_dir(src_dir: &Path, work_dir: &Path) -> io::Result<u64> {
````
Since we do not use the return value of this function, the return type should be `io::Result<()>`

2moe replied:
I’ve updated the code to make use of the `Result<u64>` return value, since I believe it’s meaningful. 
This value represents the number of bytes transferred during the copy process, and we can use it to determine whether an empty directory was copied to the `{out_dir}/mruby` path.

**In my tests:**
- Copying an empty directory returns `Ok(0)`.
- Copying a symlink `A` (pointing to folder `AA`) to `B` returns `Ok(n)`, where `n` is the size of `AA`, meaning it follows symlinks automatically.

In `copy_to_mruby_dir()`, when the **mruby** directory already exists and is non-empty, the result is `Ok(1)`. Personally, I expected a larger value, but I assume you prefer simpler behavior.

genya0407 replied:
I don't fully agree with the necessity of `u64`.
If `Ok(0)` results in `panic!`, it should be `Err` not `Ok`.

However, I will accept this code for now.

genya0407 commented on minutus-mruby-build-utils/src/mruby_dir.rs:
```diff
@@ -0,0 +1,88 @@
+        // operations.
+        fs::remove_dir_all(&target_dir)?
+    }
+    println!("cargo:warning=src dir: {src_dir:?}, target dir: {target_dir:?}");
```
I think warn is for unusual cases, and info is for usual cases.

```suggestion
    if target_dir.exists() {
        if dir_is_not_empty(&target_dir) {
            println!("cargo:info=Dir exists: {target_dir:?}");
            return Ok(0);
        }
        // An existing but empty mruby directory is a clear sign of a failed or
        // incomplete build.
        // Perform cleanup at this stage to prevent issues with subsequent `fs::rename`
        // operations.
        fs::remove_dir_all(&target_dir)?
        println!("cargo:warn=removed existing target dir: {target_dir:?}");
    }
    println!("cargo:info=src dir: {src_dir:?}, target dir: {target_dir:?}");
```

2moe replied:
Unfortunately, Cargo currently does not support `cargo:info=` or `cargo:warn=`.  
Only Cargo v1.77 and above support `cargo::error=`, but fortunately, older versions of Cargo do support `cargo:warning=`.  
For the code above, using `cargo:warning=` is appropriate—since the current stable version of Cargo does not support `cargo::debug=` or `cargo::info=`.

<https://github.com/rust-lang/cargo/issues/7037>
<https://doc.rust-lang.org/cargo/reference/build-scripts.html?highlight=warning#outputs-of-the-build-script>

2moe replied:
Although I don’t fully agree with your suggestion, I’ve still committed the related changes — just to make you happy and in hopes that this PR will be merged.

I chose `eprintln!("[INFO]")` because neither `println!("cargo:info=")` nor `println!("cargo::info=")` is supported. Since none of them are recognized by Cargo(without `-vv`) anyway, I went with the one that best matches conventional logging practices.

genya0407 replied:
Oh sorry, I thought cargo has `cargo:info=`. 

> I chose eprintln!("[INFO]")
> Since none of them are recognized by Cargo(without -vv)

I feel this is good 👍

genya0407 commented on minutus-mruby-build-utils/src/mruby_dir.rs:
```diff
@@ -0,0 +1,88 @@
+    }
+    println!("cargo:warning=src dir: {src_dir:?}, target dir: {target_dir:?}");
+
+    let status = try_to_copy_dir(src_dir, work_dir);
```
```suggestion
    let status = try_to_copy_dir(src_dir, work_dir)?;
```

2moe replied:
The `status` variable holds a `Result<u64>`, representing the number of bytes copied. I avoided an early return to ensure the function returns this value at the end. Additionally, I’ve improved the error handling to give this result more meaningful.

genya0407 commented on minutus-mruby-build-utils/src/mruby_dir.rs:
```diff
@@ -0,0 +1,88 @@
+    // Therefore, manual renaming is required.
+    //
+    // In other words:
+    // `dir::copy("/tmp/xx", "/out"); rename("/out/xx", "/out/mruby");`
```
Does `copy_inside(true)` resolve this issue?
https://docs.rs/fs_extra/latest/fs_extra/dir/struct.CopyOptions.html#structfield.copy_inside

With this, we can write `fs_extra::copy(src_dir, work_dir.join("mruby"))`, which creates `out/mruby` if it does not exist.

2moe replied:
Good catch — I totally missed that copy_inside would simplify things.

genya0407 reviewed (commented):
Thank you for your additional effort!
I added another comments.

I appreciate your detailed explanation.
I understand your point, but for this PR I would like to proceed with your current code (mutually exclusive features, condition checks without type-level guards).
https://github.com/genya0407/minutus/pull/32#issuecomment-3506092435

2moe committed:
`chore(build-utils): update error handling for copy_to_mruby_dir()`

2moe committed:
docs(mruby_dir): `fs::rename` => `dir::copy`

2moe committed:
docs(mruby_dir): add a warning before removing the dir

2moe committed:
chore(mruby_dir::try_to_copy_dir): adjust indentation

2moe committed:
chore(mruby_dir): attempted to use a log level not supported by build.rs

genya0407 reviewed (approved):
Thank you for your contribution and patiance!

genya0407 committed:
Fix README to pass CI
